### ISSUE TITLE
Some reports fail with "ERROR:  subquery in FROM must have an alias"

### ISSUE THREAD

**Issue by guitarpicker** [netdisco/netdisco#1177]
## Expected Behavior

Reports run successfully and return appropriate results.

## Current Behavior

The following reports (and possibly others) generate errors when run.
- VLANs Known but Not Configured
- VLANs No Longer Used
- VLANs Only On Uplinks

The web UI returns the message, "Search failed! Please contact your site administrator (server error)."
The Docker console shows corresponding "DBD::Pg::st execute failed: ERROR:  subquery in FROM must have an alias" errors like this:
```
netdisco-netdisco-web-1         | [20] 2024-02-28 15:27:14 error request to GET /ajax/content/report/vlansunused crashed: DBIx::Class::Storage::DBI::_dbh_execute(): DBI Exception: DBD::Pg::st execute failed: ERROR:  subquery in FROM must have an alias
netdisco-netdisco-web-1         | LINE 3: ... SELECT *, (row_number() over (partition by ip)) AS x FROM (
netdisco-netdisco-web-1         |                                                                       ^
netdisco-netdisco-web-1         | HINT:  For example, FROM (SELECT ...) [AS] foo. [for Statement "SELECT me.ip, me.vlans FROM (SELECT ip, array_agg(vlans) AS vlans FROM (
netdisco-netdisco-web-1         |   SELECT ip, array_to_string(array_agg(vlan), ', ') AS vlans, (x / COALESCE(NULLIF(?,''), '20') ::integer) AS chunk FROM (
netdisco-netdisco-web-1         |     SELECT *, (row_number() over (partition by ip)) AS x FROM (
netdisco-netdisco-web-1         |
netdisco-netdisco-web-1         |       SELECT DISTINCT ip, vlan
netdisco-netdisco-web-1         |         FROM device_port_vlan dpv
netdisco-netdisco-web-1         |         WHERE dpv.native IS false
netdisco-netdisco-web-1         |           AND dpv.vlan <> 1
netdisco-netdisco-web-1         |           AND (
netdisco-netdisco-web-1         |             SELECT count(*) FROM device_port_vlan dpv2
netdisco-netdisco-web-1         |               LEFT JOIN device_port dp USING (ip, port)
netdisco-netdisco-web-1         |               LEFT JOIN device d USING (ip)
netdisco-netdisco-web-1         |               WHERE dpv2.ip = dpv.ip
netdisco-netdisco-web-1         |                 AND dpv2.vlan = dpv.vlan
netdisco-netdisco-web-1         |                 AND native IS true
netdisco-netdisco-web-1         |                 AND (
netdisco-netdisco-web-1         |                   dp.up_admin = 'up'
netdisco-netdisco-web-1         |                   OR age( LOCALTIMESTAMP,
netdisco-netdisco-web-1         |                           to_timestamp( extract( epoch FROM d.last_discover ) - ( d.uptime - dp.lastchange ) /100 ) ::timestamp )
netdisco-netdisco-web-1         |                     < COALESCE(NULLIF(?,''), '3 months') ::interval
netdisco-netdisco-web-1         |                 )
netdisco-netdisco-web-1         |           ) = 0
netdisco-netdisco-web-1         |         ORDER BY dpv.ip, dpv.vlan)
netdisco-netdisco-web-1         |
netdisco-netdisco-web-1         |   ) GROUP BY ip, chunk
netdisco-netdisco-web-1         | ) GROUP BY ip ORDER BY ip
netdisco-netdisco-web-1         | ) me" with ParamValues: 1='', 2=''] at /home/netdisco/perl5/lib/perl5/App/Netdisco/Web/GenericReport.pm line 66
```

## Possible Solution

I'm not an expert, but it appears that Postgresql requires that subqueries have names.  Add some subquery names to the query (preferred), or use a version of Postgresql that does not have this requirement.

This article discusses the issue somewhat.  It appears that newer versions of Postgresql might not have this limitation. 
 https://pganalyze.com/blog/5mins-postgres-waiting-for-postgres-16-subquery-alias-optional

For example, I ran the following query taken from [config.yml](https://github.com/netdisco/netdisco/blob/5fae76d808858da838cfc356004b8ddc6abc94b4/share/config.yml#L4), replacing the ? placeholder with '' and adding xxx, yyy,and zzz aliases.  After making these modifications, it returned the results without an error.
```
SELECT ip, array_agg(vlans) AS vlans FROM (
        SELECT ip, array_to_string(array_agg(vlan), ', ') AS vlans, (x / COALESCE(NULLIF('',''), '20') ::integer) AS chunk FROM (
          SELECT *, (row_number() over (partition by ip)) AS x FROM (

            SELECT DISTINCT ip, vlan
              FROM device_port_vlan dpv
              WHERE dpv.native IS false
                AND dpv.vlan <> 1
                AND (
                  SELECT count(*) FROM device_port_vlan dpv2
                    LEFT JOIN device_port dp USING (ip, port)
                    LEFT JOIN device d USING (ip)
                    WHERE dpv2.ip = dpv.ip
                      AND dpv2.vlan = dpv.vlan
                      AND native IS true
                      AND (
                        dp.up_admin = 'up'
                        OR age( LOCALTIMESTAMP,
                                to_timestamp( extract( epoch FROM d.last_discover ) - ( d.uptime - dp.lastchange ) /100 ) ::timestamp )
                          < COALESCE(NULLIF('',''), '3 months') ::interval
                      ) 
                ) = 0
              ORDER BY dpv.ip, dpv.vlan) AS zzz

        ) as yyy GROUP BY ip, chunk
      ) as xxx GROUP BY ip ORDER BY ip;
```

## Steps to Reproduce (for bugs)

1. Navigate to the Reports --> VLAN --> VLANs no longer used.

## Context

I am unable to run the affected reports.

## Your Environment

* netdisco container versions: 
```
~/projects/netdisco/netdisco (master *)# docker compose ps
NAME                             IMAGE                                 COMMAND                                               SERVICE               CREATED        STATUS        PORTS
netdisco-netdisco-backend-1      netdisco/netdisco:latest-backend      "/home/netdisco/bin/netdisco-backend foreground"      netdisco-backend      16 hours ago   Up 16 hours
netdisco-netdisco-postgresql-1   netdisco/netdisco:latest-postgresql   "/usr/local/bin/netdisco-db-entrypoint.sh postgres"   netdisco-postgresql   16 hours ago   Up 16 hours   5432/tcp
netdisco-netdisco-web-1          netdisco/netdisco:latest-web          "/home/netdisco/bin/netdisco-web foreground"          netdisco-web          16 hours ago   Up 16 hours   0.0.0.0:5000->5000/tcp

# docker image ls | grep netdisco.*latest
netdisco/netdisco            latest-backend      71083b247872   13 days ago     640MB
netdisco/netdisco            latest-web          54b410dce356   13 days ago     130MB
netdisco/netdisco            latest-postgresql   7d707403e704   13 days ago     199MB
netdisco/netdisco            latest-do           16ab00f04632   21 months ago   496MB
```
* Postgres version: PostgreSQL 13.4 on x86_64-pc-linux-musl, compiled by gcc (Alpine 10.3.1_git20210424) 10.3.1 20210424, 64-bit
* docker engine version: Docker version 24.0.7, build afdd53b
* docker-compose version: 2.3
* host operating system: Red Hat Enterprise Linux release 8.9 (Ootpa)

## Config info (deployment.yml and docker env settings)
I have a vanilla deployment.yml plus customized SNMP credentials.

## Device information
I have an assortment of mainly HP/Aruba and Cisco switches.

**Comment by rc9000** [comment]
This should fix it, tested on Postgres 12.17.

**Comment by guitarpicker** [comment]
You all are the best.  I just pulled a fresh Docker image, and those reports are working now... and providing some useful results!
